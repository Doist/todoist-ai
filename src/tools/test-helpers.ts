import type { PersonalProject, Section, Task } from '@doist/todoist-api-typescript'

/**
 * Mapped task type matching the output of mapTask function.
 * Used for filter-based query test mocks.
 */
export type MappedTask = {
    id: string
    content: string
    description: string
    dueDate: string | undefined
    recurring: string | boolean
    priority: number
    projectId: string
    sectionId: string | null
    parentId: string | null
    labels: string[]
}

/**
 * Creates a mock Task with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockTask(overrides: Partial<Task> = {}): Task {
    return {
        id: '8485093748',
        content: 'Test task content',
        description: '',
        completedAt: null,
        labels: [],
        childOrder: 1,
        priority: 1,
        projectId: '6cfCcrrCFg2xP94Q',
        sectionId: null,
        parentId: null,
        url: 'https://todoist.com/showTask?id=8485093748',
        // Use correct property names from Task schema
        noteCount: 0,
        addedByUid: '713437',
        addedAt: '2025-08-13T22:09:56.123456Z',
        deadline: null,
        responsibleUid: null,
        assignedByUid: null,
        isCollapsed: false,
        isDeleted: false,
        duration: null,
        checked: false,
        updatedAt: '2025-08-13T22:09:56.123456Z',
        due: null,
        dayOrder: 0,
        userId: '713437',
        ...overrides,
    }
}

/**
 * Creates a mock Section with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockSection(overrides: Partial<Section> = {}): Section {
    return {
        id: 'section-123',
        projectId: '6cfCcrrCFg2xP94Q',
        sectionOrder: 1,
        userId: 'test-user',
        addedAt: '2024-01-01T00:00:00Z',
        updatedAt: '2024-01-01T00:00:00Z',
        archivedAt: null,
        isArchived: false,
        isDeleted: false,
        isCollapsed: false,
        name: 'Test Section',
        ...overrides,
    }
}

/**
 * Creates a mock PersonalProject with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockProject(overrides: Partial<PersonalProject> = {}): PersonalProject {
    return {
        id: '6cfCcrrCFg2xP94Q',
        name: 'Test Project',
        color: 'charcoal',
        isFavorite: false,
        isShared: false,
        parentId: null,
        inboxProject: false,
        viewStyle: 'list',
        url: 'https://todoist.com/projects/6cfCcrrCFg2xP94Q',
        isDeleted: false,
        updatedAt: '2025-08-13T22:09:55.841800Z',
        createdAt: '2025-08-13T22:09:55.841785Z',
        childOrder: 1,
        defaultOrder: 0,
        description: '',
        isCollapsed: false,
        canAssignTasks: false,
        isFrozen: false,
        isArchived: false,
        ...overrides,
    }
}

/**
 * Creates a mock API response object with results and nextCursor.
 */
export function createMockApiResponse<T>(
    results: T[],
    nextCursor: string | null = null,
): {
    results: T[]
    nextCursor: string | null
} {
    return {
        results,
        nextCursor,
    }
}

/**
 * Creates a simplified mapped task (matches mapTask output) for filter-based query tests.
 */
export function createMappedTask(overrides: Partial<MappedTask> = {}): MappedTask {
    return {
        id: TEST_IDS.TASK_1,
        content: 'Test task content',
        description: '',
        dueDate: undefined,
        recurring: false,
        priority: 1,
        projectId: TEST_IDS.PROJECT_TEST,
        sectionId: null,
        parentId: null,
        labels: [],
        ...overrides,
    }
}

/**
 * Common error messages used across tests.
 */
export const TEST_ERRORS = {
    API_RATE_LIMIT: 'API Error: Rate limit exceeded',
    API_UNAUTHORIZED: 'API Error: Unauthorized',
    INVALID_CURSOR: 'Invalid cursor format',
    INVALID_FILTER: 'Invalid filter query',
} as const

/**
 * Creates multiple test cases for parameterized testing.
 */
export function createTestCases<T, E = unknown>(
    cases: Array<{ name: string; input: T; expected?: E }>,
) {
    return cases
}

/**
 * Common mock IDs used across tests for consistency.
 */
export const TEST_IDS = {
    TASK_1: '8485093748',
    TASK_2: '8485093749',
    TASK_3: '8485093750',
    PROJECT_INBOX: 'inbox-project-id',
    PROJECT_WORK: 'work-project-id',
    PROJECT_TEST: '6cfCcrrCFg2xP94Q',
    SECTION_1: 'section-123',
    SECTION_2: 'section-456',
    USER_ID: '713437',
} as const
