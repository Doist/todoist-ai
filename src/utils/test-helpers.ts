import type { PersonalProject, Section, Task } from '@doist/todoist-api-typescript'
import { getToolOutput } from '../mcp-helpers'

/**
 * Mapped task type matching the output of mapTask function.
 * Used for filter-based query test mocks.
 */
export type MappedTask = {
    id: string
    content: string
    description: string
    dueDate: string | undefined
    recurring: string | boolean
    priority: number
    projectId: string
    sectionId: string | null
    parentId: string | null
    labels: string[]
    duration: string | null
    responsibleUid: string | null
    assignedByUid: string | null
}

/**
 * Creates a mock Task with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockTask(overrides: Partial<Task> = {}): Task {
    return {
        id: '8485093748',
        content: 'Test task content',
        description: '',
        completedAt: null,
        labels: [],
        childOrder: 1,
        priority: 1,
        projectId: '6cfCcrrCFg2xP94Q',
        sectionId: null,
        parentId: null,
        url: 'https://todoist.com/showTask?id=8485093748',
        // Use correct property names from Task schema
        noteCount: 0,
        addedByUid: '713437',
        addedAt: '2025-08-13T22:09:56.123456Z',
        deadline: null,
        responsibleUid: null,
        assignedByUid: null,
        isCollapsed: false,
        isDeleted: false,
        duration: null,
        checked: false,
        updatedAt: '2025-08-13T22:09:56.123456Z',
        due: null,
        dayOrder: 0,
        userId: '713437',
        ...overrides,
    }
}

/**
 * Creates a mock Section with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockSection(overrides: Partial<Section> = {}): Section {
    return {
        id: 'section-123',
        projectId: '6cfCcrrCFg2xP94Q',
        sectionOrder: 1,
        userId: 'test-user',
        addedAt: '2024-01-01T00:00:00Z',
        updatedAt: '2024-01-01T00:00:00Z',
        archivedAt: null,
        isArchived: false,
        isDeleted: false,
        isCollapsed: false,
        name: 'Test Section',
        url: 'https://todoist.com/sections/section-123',
        ...overrides,
    }
}

/**
 * Creates a mock PersonalProject with all required properties and sensible defaults.
 * Pass only the properties you want to override for your specific test.
 */
export function createMockProject(overrides: Partial<PersonalProject> = {}): PersonalProject {
    return {
        id: '6cfCcrrCFg2xP94Q',
        name: 'Test Project',
        color: 'charcoal',
        isFavorite: false,
        isShared: false,
        parentId: null,
        inboxProject: false,
        viewStyle: 'list',
        url: 'https://todoist.com/projects/6cfCcrrCFg2xP94Q',
        isDeleted: false,
        updatedAt: '2025-08-13T22:09:55.841800Z',
        createdAt: '2025-08-13T22:09:55.841785Z',
        childOrder: 1,
        defaultOrder: 0,
        description: '',
        isCollapsed: false,
        canAssignTasks: false,
        isFrozen: false,
        isArchived: false,
        ...overrides,
    }
}

/**
 * Creates a mock API response object with results and nextCursor.
 */
export function createMockApiResponse<T>(
    results: T[],
    nextCursor: string | null = null,
): {
    results: T[]
    nextCursor: string | null
} {
    return {
        results,
        nextCursor,
    }
}

/**
 * Creates a simplified mapped task (matches mapTask output) for filter-based query tests.
 */
export function createMappedTask(overrides: Partial<MappedTask> = {}): MappedTask {
    return {
        id: TEST_IDS.TASK_1,
        content: 'Test task content',
        description: '',
        dueDate: undefined,
        recurring: false,
        priority: 1,
        projectId: TEST_IDS.PROJECT_TEST,
        sectionId: null,
        parentId: null,
        labels: [],
        duration: null,
        responsibleUid: null,
        assignedByUid: null,
        ...overrides,
    }
}

/**
 * Common error messages used across tests.
 */
export const TEST_ERRORS = {
    API_RATE_LIMIT: 'API Error: Rate limit exceeded',
    API_UNAUTHORIZED: 'API Error: Unauthorized',
    INVALID_CURSOR: 'Invalid cursor format',
    INVALID_FILTER: 'Invalid filter query',
} as const

/**
 * Creates multiple test cases for parameterized testing.
 */
export function createTestCases<T, E = unknown>(
    cases: Array<{ name: string; input: T; expected?: E }>,
) {
    return cases
}

/**
 * Extracts the text content from a tool output for snapshot testing.
 * This allows tests to match against just the text content while tools return structured output.
 */
export function extractTextContent(toolOutput: unknown): string {
    if (typeof toolOutput === 'string') {
        return toolOutput
    }

    if (typeof toolOutput === 'object' && toolOutput !== null && 'content' in toolOutput) {
        const output = toolOutput as { content: unknown }
        if (
            Array.isArray(output.content) &&
            output.content[0] &&
            typeof output.content[0] === 'object' &&
            output.content[0] !== null &&
            'type' in output.content[0] &&
            'text' in output.content[0] &&
            output.content[0].type === 'text'
        ) {
            return output.content[0].text as string
        }
    }

    throw new Error('Expected tool output to have text content')
}

/**
 * Extracts the structured content from a tool output for testing.
 * This handles both the new `structuredContent` field and legacy JSON-encoded content.
 */
export function extractStructuredContent(
    output: ReturnType<typeof getToolOutput>,
): Record<string, unknown> {
    // Check for new structuredContent field first
    if ('structuredContent' in output && typeof output.structuredContent === 'object') {
        return output.structuredContent as Record<string, unknown>
    }

    // Fall back to checking for JSON content in the content array
    if ('content' in output && Array.isArray(output.content)) {
        for (const item of output.content) {
            if (
                typeof item === 'object' &&
                item !== null &&
                'type' in item &&
                'text' in item &&
                item.type === 'text' &&
                item.mimeType === 'application/json'
            ) {
                return JSON.parse(item.text) as Record<string, unknown>
            }
        }
    }

    throw new Error('Expected tool output to have structured content')
}

/**
 * Common mock IDs used across tests for consistency.
 */
export const TEST_IDS = {
    TASK_1: '8485093748',
    TASK_2: '8485093749',
    TASK_3: '8485093750',
    PROJECT_INBOX: 'inbox-project-id',
    PROJECT_WORK: 'work-project-id',
    PROJECT_TEST: '6cfCcrrCFg2xP94Q',
    SECTION_1: 'section-123',
    SECTION_2: 'section-456',
    USER_ID: '713437',
} as const

/**
 * Fixed date for consistent test snapshots.
 * Use this instead of new Date() in tests to avoid snapshot drift.
 */
export const TODAY = '2025-08-17' as const
