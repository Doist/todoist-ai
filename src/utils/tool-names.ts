/**
 * Centralized tool names module
 *
 * This module provides a single source of truth for all tool names used throughout the codebase.
 * Each tool should import its own name from this module to avoid hardcoded strings.
 * This prevents outdated references when tool names change.
 */

export const TOOL_NAMES = {
    // Task management tools
    TASKS_ADD_MULTIPLE: 'tasks-add-multiple',
    TASKS_COMPLETE_MULTIPLE: 'tasks-complete-multiple',
    TASKS_UPDATE_MULTIPLE: 'tasks-update-multiple',
    TASKS_SEARCH: 'tasks-search',
    TASKS_LIST_BY_DATE: 'tasks-list-by-date',
    TASKS_LIST_COMPLETED: 'tasks-list-completed',
    TASKS_LIST_FOR_CONTAINER: 'tasks-list-for-container',

    // Project management tools
    PROJECTS_LIST: 'projects-list',
    PROJECTS_MANAGE: 'projects-manage',

    // Section management tools
    SECTIONS_SEARCH: 'sections-search',
    SECTIONS_MANAGE: 'sections-manage',

    // General tools
    OVERVIEW: 'overview',
    DELETE_ONE: 'delete-one',
} as const

// Type for all tool names
export type ToolName = (typeof TOOL_NAMES)[keyof typeof TOOL_NAMES]
