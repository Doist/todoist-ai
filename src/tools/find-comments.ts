import type { Comment } from '@doist/todoist-api-typescript'
import { z } from 'zod'
import { getToolOutput } from '../mcp-helpers.js'
import type { TodoistTool } from '../todoist-tool.js'
import { ApiLimits } from '../utils/constants.js'
import { formatNextSteps } from '../utils/response-builders.js'
import { ToolNames } from '../utils/tool-names.js'

const { ADD_COMMENTS, UPDATE_COMMENTS, DELETE_OBJECT } = ToolNames

const ArgsSchema = {
    taskId: z.string().optional().describe('Find comments for a specific task.'),
    projectId: z.string().optional().describe('Find comments for a specific project.'),
    commentId: z.string().optional().describe('Get a specific comment by ID.'),
    cursor: z.string().optional().describe('Pagination cursor for retrieving more results.'),
    limit: z
        .number()
        .int()
        .min(1)
        .max(ApiLimits.COMMENTS_MAX)
        .optional()
        .describe('Maximum number of comments to return'),
}

const findComments = {
    name: ToolNames.FIND_COMMENTS,
    description:
        'Find comments by task, project, or get a specific comment by ID. Exactly one of taskId, projectId, or commentId must be provided.',
    parameters: ArgsSchema,
    async execute(args, client) {
        // Validate that exactly one search parameter is provided
        const searchParams = [args.taskId, args.projectId, args.commentId].filter(Boolean)
        if (searchParams.length === 0) {
            throw new Error('Must provide exactly one of: taskId, projectId, or commentId.')
        }
        if (searchParams.length > 1) {
            throw new Error(
                'Cannot provide multiple search parameters. Choose one of: taskId, projectId, or commentId.',
            )
        }

        let comments: Comment[]
        let hasMore = false
        let nextCursor: string | null = null

        if (args.commentId) {
            // Get single comment
            const comment = await client.getComment(args.commentId)
            comments = [comment]
        } else if (args.taskId) {
            // Get comments by task
            const response = await client.getComments({
                taskId: args.taskId,
                cursor: args.cursor || null,
                limit: args.limit || ApiLimits.COMMENTS_DEFAULT,
            })
            comments = response.results
            hasMore = response.nextCursor !== null
            nextCursor = response.nextCursor
        } else if (args.projectId) {
            // Get comments by project
            const response = await client.getComments({
                projectId: args.projectId,
                cursor: args.cursor || null,
                limit: args.limit || ApiLimits.COMMENTS_DEFAULT,
            })
            comments = response.results
            hasMore = response.nextCursor !== null
            nextCursor = response.nextCursor
        } else {
            // This should never happen due to validation, but TypeScript needs it
            throw new Error('Invalid state: no search parameter provided')
        }

        const textContent = generateTextContent({
            comments,
            searchType: args.commentId ? 'single' : args.taskId ? 'task' : 'project',
            searchId: args.commentId || args.taskId || args.projectId || '',
            hasMore,
            nextCursor,
        })

        return getToolOutput({
            textContent,
            structuredContent: {
                comments,
                searchType: args.commentId ? 'single' : args.taskId ? 'task' : 'project',
                searchId: args.commentId || args.taskId || args.projectId || '',
                hasMore,
                nextCursor,
                totalCount: comments.length,
            },
        })
    },
} satisfies TodoistTool<typeof ArgsSchema>

function generateTextContent({
    comments,
    searchType,
    searchId,
    hasMore,
    nextCursor,
}: {
    comments: Comment[]
    searchType: 'single' | 'task' | 'project'
    searchId: string
    hasMore: boolean
    nextCursor: string | null
}): string {
    if (comments.length === 0) {
        return `No comments found for ${searchType}${searchType !== 'single' ? ` ${searchId}` : ''}`
    }

    // Build summary
    let summary: string
    if (searchType === 'single') {
        const comment = comments[0]
        if (!comment) {
            return 'Comment not found'
        }
        const hasAttachment = comment.fileAttachment !== null
        const attachmentInfo = hasAttachment
            ? ` • Has attachment: ${comment.fileAttachment?.fileName || 'file'}`
            : ''
        summary = `Found comment${attachmentInfo} • id=${comment.id}`
    } else {
        const attachmentCount = comments.filter((c) => c.fileAttachment !== null).length
        const attachmentInfo = attachmentCount > 0 ? ` (${attachmentCount} with attachments)` : ''
        const commentsLabel = comments.length === 1 ? 'comment' : 'comments'
        summary = `Found ${comments.length} ${commentsLabel} for ${searchType} ${searchId}${attachmentInfo}`

        if (hasMore) {
            summary += ' • More available'
        }
    }

    // Context-aware next steps
    const nextSteps: string[] = []

    if (searchType === 'single') {
        const comment = comments[0]
        if (comment) {
            nextSteps.push(`Use ${UPDATE_COMMENTS} with id=${comment.id} to edit content`)
            nextSteps.push(`Use ${DELETE_OBJECT} with type=comment id=${comment.id} to remove`)

            // Suggest viewing related comments
            if (comment.taskId) {
                nextSteps.push(
                    `Use ${ToolNames.FIND_COMMENTS} with taskId=${comment.taskId} to see all task comments`,
                )
            } else if (comment.projectId) {
                nextSteps.push(
                    `Use ${ToolNames.FIND_COMMENTS} with projectId=${comment.projectId} to see all project comments`,
                )
            }
        }
    } else {
        // Multiple comments
        nextSteps.push(`Use ${ADD_COMMENTS} with ${searchType}Id=${searchId} to add new comment`)

        if (comments.length > 0) {
            nextSteps.push(
                `Use ${ToolNames.FIND_COMMENTS} with commentId to view specific comment details`,
            )
        }

        // Pagination
        if (hasMore && nextCursor) {
            nextSteps.push(
                `Use ${ToolNames.FIND_COMMENTS} with cursor="${nextCursor}" to get more results`,
            )
        }
    }

    const next = formatNextSteps(nextSteps)
    return `${summary}\n${next}`
}

export { findComments }
