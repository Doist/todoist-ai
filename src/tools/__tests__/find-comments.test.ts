import type { Comment, TodoistApi } from '@doist/todoist-api-typescript'
import { jest } from '@jest/globals'
import { extractStructuredContent, extractTextContent } from '../../utils/test-helpers.js'
import { ToolNames } from '../../utils/tool-names.js'
import { findComments } from '../find-comments.js'

// Mock the Todoist API
const mockTodoistApi = {
    getComment: jest.fn(),
    getComments: jest.fn(),
} as unknown as jest.Mocked<TodoistApi>

const { FIND_COMMENTS } = ToolNames

const createMockComment = (overrides: Partial<Comment> = {}): Comment => ({
    id: '12345',
    content: 'Test comment content',
    postedAt: '2024-01-01T12:00:00Z',
    postedUid: 'user123',
    taskId: 'task123',
    projectId: undefined,
    fileAttachment: null,
    uidsToNotify: null,
    reactions: null,
    isDeleted: false,
    ...overrides,
})

describe(`${FIND_COMMENTS} tool`, () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    describe('finding comments by task', () => {
        it('should find comments for a task', async () => {
            const mockComments = [
                createMockComment({ id: '1', content: 'First comment', taskId: 'task123' }),
                createMockComment({ id: '2', content: 'Second comment', taskId: 'task123' }),
            ]

            mockTodoistApi.getComments.mockResolvedValue({
                results: mockComments,
                nextCursor: null,
            })

            const result = await findComments.execute(
                {
                    taskId: 'task123',
                },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getComments).toHaveBeenCalledWith({
                taskId: 'task123',
                cursor: null,
                limit: 10,
            })

            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    comments: expect.arrayContaining([
                        expect.objectContaining({ id: '1', content: 'First comment' }),
                        expect.objectContaining({ id: '2', content: 'Second comment' }),
                    ]),
                    searchType: 'task',
                    searchId: 'task123',
                    hasMore: false,
                    nextCursor: null,
                    totalCount: 2,
                }),
            )
        })

        it('should handle pagination', async () => {
            const mockComments = [createMockComment({ id: '1', content: 'Comment 1' })]

            mockTodoistApi.getComments.mockResolvedValue({
                results: mockComments,
                nextCursor: 'next_page_token',
            })

            const result = await findComments.execute(
                {
                    taskId: 'task123',
                    limit: 1,
                    cursor: 'current_cursor',
                },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getComments).toHaveBeenCalledWith({
                taskId: 'task123',
                cursor: 'current_cursor',
                limit: 1,
            })

            // Verify result includes pagination info
            expect(extractTextContent(result)).toMatchSnapshot()

            // Verify structured content includes pagination
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    comments: expect.arrayContaining([
                        expect.objectContaining({ id: '1', content: 'Comment 1' }),
                    ]),
                    searchType: 'task',
                    searchId: 'task123',
                    hasMore: true,
                    nextCursor: 'next_page_token',
                    totalCount: 1,
                }),
            )
        })
    })

    describe('finding comments by project', () => {
        it('should find comments for a project', async () => {
            const mockComments = [
                createMockComment({
                    id: '1',
                    content: 'Project comment',
                    taskId: undefined,
                    projectId: 'project456',
                }),
            ]

            mockTodoistApi.getComments.mockResolvedValue({
                results: mockComments,
                nextCursor: null,
            })

            const result = await findComments.execute(
                {
                    projectId: 'project456',
                },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getComments).toHaveBeenCalledWith({
                projectId: 'project456',
                cursor: null,
                limit: 10,
            })

            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    comments: expect.arrayContaining([
                        expect.objectContaining({
                            id: '1',
                            content: 'Project comment',
                            projectId: 'project456',
                        }),
                    ]),
                    searchType: 'project',
                    searchId: 'project456',
                    hasMore: false,
                    nextCursor: null,
                    totalCount: 1,
                }),
            )
        })
    })

    describe('finding single comment', () => {
        it('should find comment by ID', async () => {
            const mockComment = createMockComment({
                id: 'comment789',
                content: 'Single comment content',
                taskId: 'task123',
            })

            mockTodoistApi.getComment.mockResolvedValue(mockComment)

            const result = await findComments.execute(
                {
                    commentId: 'comment789',
                },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getComment).toHaveBeenCalledWith('comment789')

            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    comments: expect.arrayContaining([
                        expect.objectContaining({
                            id: 'comment789',
                            content: 'Single comment content',
                            taskId: 'task123',
                            fileAttachment: null,
                        }),
                    ]),
                    searchType: 'single',
                    searchId: 'comment789',
                    hasMore: false,
                    nextCursor: null,
                    totalCount: 1,
                }),
            )
        })

        it('should handle comment with attachment', async () => {
            const mockComment = createMockComment({
                id: 'comment789',
                content: 'Comment with file',
                fileAttachment: {
                    resourceType: 'file',
                    fileName: 'document.pdf',
                    fileUrl: 'https://example.com/document.pdf',
                    fileType: 'application/pdf',
                },
            })

            mockTodoistApi.getComment.mockResolvedValue(mockComment)

            const result = await findComments.execute(
                {
                    commentId: 'comment789',
                },
                mockTodoistApi,
            )

            // Verify result includes attachment info
            expect(extractTextContent(result)).toMatchSnapshot()

            // Verify structured content includes attachment
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    comments: expect.arrayContaining([
                        expect.objectContaining({
                            id: 'comment789',
                            content: 'Comment with file',
                            fileAttachment: expect.objectContaining({
                                resourceType: 'file',
                                fileName: 'document.pdf',
                                fileUrl: 'https://example.com/document.pdf',
                                fileType: 'application/pdf',
                            }),
                        }),
                    ]),
                    searchType: 'single',
                    searchId: 'comment789',
                    hasMore: false,
                    nextCursor: null,
                    totalCount: 1,
                }),
            )
        })
    })

    describe('validation', () => {
        it('should throw error when no search parameter provided', async () => {
            await expect(findComments.execute({}, mockTodoistApi)).rejects.toThrow(
                'Must provide exactly one of: taskId, projectId, or commentId.',
            )
        })

        it('should throw error when multiple search parameters provided', async () => {
            await expect(
                findComments.execute(
                    {
                        taskId: 'task123',
                        projectId: 'project456',
                    },
                    mockTodoistApi,
                ),
            ).rejects.toThrow('Cannot provide multiple search parameters.')
        })
    })

    describe('empty results', () => {
        it('should handle no comments found', async () => {
            mockTodoistApi.getComments.mockResolvedValue({
                results: [],
                nextCursor: null,
            })

            const result = await findComments.execute(
                {
                    taskId: 'task123',
                },
                mockTodoistApi,
            )

            // Verify result handles empty case
            expect(extractTextContent(result)).toMatchSnapshot()

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    comments: [],
                    searchType: 'task',
                    searchId: 'task123',
                    hasMore: false,
                    nextCursor: null,
                    totalCount: 0,
                }),
            )
        })
    })
})
