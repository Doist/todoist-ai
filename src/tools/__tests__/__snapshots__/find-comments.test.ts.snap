// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`find-comments tool empty results should handle no comments found 1`] = `"No comments found for task task123"`;

exports[`find-comments tool finding comments by project should find comments for a project 1`] = `
"Found 1 comment for project project456
Next:
- Use add-comments with projectId=project456 to add new comment
- Use find-comments with commentId to view specific comment details"
`;

exports[`find-comments tool finding comments by task should find comments for a task 1`] = `
"Found 2 comments for task task123
Next:
- Use add-comments with taskId=task123 to add new comment
- Use find-comments with commentId to view specific comment details"
`;

exports[`find-comments tool finding comments by task should handle pagination 1`] = `
"Found 1 comment for task task123 • More available
Next:
- Use add-comments with taskId=task123 to add new comment
- Use find-comments with commentId to view specific comment details
- Use find-comments with cursor="next_page_token" to get more results"
`;

exports[`find-comments tool finding single comment should find comment by ID 1`] = `
"Found comment • id=comment789
Next:
- Use update-comments with id=comment789 to edit content
- Use delete-object with type=comment id=comment789 to remove
- Use find-comments with taskId=task123 to see all task comments"
`;

exports[`find-comments tool finding single comment should handle comment with attachment 1`] = `
"Found comment • Has attachment: document.pdf • id=comment789
Next:
- Use update-comments with id=comment789 to edit content
- Use delete-object with type=comment id=comment789 to remove
- Use find-comments with taskId=task123 to see all task comments"
`;
